crates/sui-core/src/histogram.rs

```diff
@@ -0,0 +1,402 @@
+impl HistogramVec {
+    pub fn new_in_registry(name: &str, desc: &str, labels: &[&str], registry: &Registry) -> Self {
+        Self::new_in_registry_with_percentiles(
+            name,
+            desc,
+            labels,
+            registry,
+            vec![500usize, 950, 990],
+        )
+    }
+
+    /// Allows to specify percentiles in 1/1000th, e.g. 90pct is specified as 900
+    pub fn new_in_registry_with_percentiles(
+        name: &str,
+        desc: &str,
+        labels: &[&str],
+        registry: &Registry,
+        percentiles: Vec<usize>,
+    ) -> Self {
+        let sum_name = format!("{}_sum", name);
+        let count_name = format!("{}_count", name);
+        let sum =
+            register_int_counter_vec_with_registry!(sum_name, desc, labels, registry).unwrap();
+        let count =
+            register_int_counter_vec_with_registry!(count_name, desc, labels, registry).unwrap();
+        let labels: Vec<_> = labels.iter().cloned().chain(["pct"].into_iter()).collect();
+        let gauge = register_int_gauge_vec_with_registry!(name, desc, &labels, registry).unwrap();
+        Self::new(gauge, sum, count, percentiles)
+    }
+
+    // Do not expose it to public interface because we need labels to have a specific format (e.g. add last label is "pct")
+    fn new(
+        gauge: IntGaugeVec,
+        sum: IntCounterVec,
+        count: IntCounterVec,
+        percentiles: Vec<usize>,
+    ) -> Self {
+        let (sender, receiver) = mpsc::channel(1000);
+        let reporter = HistogramReporter {
+            gauge,
+            sum,
+            count,
+            percentiles,
+            known_labels: Default::default(),
+        };
+        let reporter = Arc::new(Mutex::new(reporter));
+        let collector = HistogramCollector {
+            reporter,
+            channel: receiver,
+        };
+        Handle::current().spawn(collector.run());
+        Self { channel: sender }
+    }
+
+    pub fn with_label_values(&self, labels: &[&str]) -> Histogram {
+        let labels = labels.iter().map(ToString::to_string).collect();
+        let labels = HistogramLabelsInner::new(labels);
+        Histogram {
+            labels,
+            channel: self.channel.clone(),
+        }
+    }
+}
+
+impl HistogramLabelsInner {
+    pub fn new(labels: Vec<String>) -> HistogramLabels {
+        // Not a crypto hash
+        let mut hasher = DefaultHasher::new();
+        labels.hash(&mut hasher);
+        let hash = hasher.finish();
+        Arc::new(Self { labels, hash })
+    }
+}
+
+impl PartialEq for HistogramLabelsInner {
+    fn eq(&self, other: &Self) -> bool {
+        self.hash == other.hash
+    }
+}
+
+impl Eq for HistogramLabelsInner {}
+
+impl Hash for HistogramLabelsInner {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.hash.hash(state)
+    }
+}
+
+impl Histogram {
+    pub fn report(&self, v: Point) {
+        match self.channel.try_send((self.labels.clone(), v)) {
```

Comment:
Can we not clone the labels on each report, but instead send a u64 ID or something?   That means several allocations totalling 500 bytes - 1K per point, which is slow and also memory hungry.

Reply:
Label is an Arc so it is essentially sending `u64`